File: frontend/lib/features/research_map/data/dto/creator_pack.dart
import '../api_exception.dart';

enum CreatorAudience { generalPublic, highSchool, undergraduate }

String audienceValue(CreatorAudience value) => switch (value) {
      CreatorAudience.generalPublic => 'general_public',
      CreatorAudience.highSchool => 'high_school',
      CreatorAudience.undergraduate => 'undergraduate',
    };

class CreatorPack {
  const CreatorPack({
    required this.packId,
    required this.paperId,
    required this.audience,
    required this.status,
    required this.title,
    required this.summary,
    required this.narrationScript,
    required this.visualAbstract,
    required this.evidenceCards,
    required this.limitations,
    required this.disclaimer,
  });

  final String packId;
  final String paperId;
  final String audience;
  final String status;
  final String title;
  final String summary;
  final String narrationScript;
  final List<CreatorBlock> visualAbstract;
  final List<CreatorEvidenceCard> evidenceCards;
  final List<String> limitations;
  final String disclaimer;

  factory CreatorPack.fromJson(Map<String, Object?> json) => CreatorPack(
        packId: _text(json['pack_id']),
        paperId: _text(json['paper_id']),
        audience: _text(json['audience']),
        status: _text(json['status']),
        title: _text(json['title']),
        summary: _text(json['summary']),
        narrationScript: _text(json['narration_script']),
        visualAbstract: _objects(json['visual_abstract'])
            .map(CreatorBlock.fromJson)
            .toList(),
        evidenceCards: _objects(json['evidence_cards'])
            .map(CreatorEvidenceCard.fromJson)
            .toList(),
        limitations: _strings(json['limitations']),
        disclaimer: _text(json['disclaimer']),
      );
}

class CreatorBlock {
  const CreatorBlock({required this.label, required this.text});
  final String label;
  final String text;
  factory CreatorBlock.fromJson(Map<String, Object?> json) =>
      CreatorBlock(label: _text(json['label']), text: _text(json['text']));
}

class CreatorEvidenceCard {
  const CreatorEvidenceCard({
    required this.statement,
    required this.confidence,
    required this.evidenceIds,
  });
  final String statement;
  final String confidence;
  final List<String> evidenceIds;
  factory CreatorEvidenceCard.fromJson(Map<String, Object?> json) =>
      CreatorEvidenceCard(
        statement: _text(json['statement']),
        confidence: _text(json['confidence']),
        evidenceIds: _strings(json['evidence_ids']),
      );
}

String _text(Object? value) {
  if (value is! String || value.trim().isEmpty) throw const ParseException();
  return value.trim();
}

List<Map<String, Object?>> _objects(Object? value) {
  if (value is! List) throw const ParseException();
  return value.map((item) {
    if (item is! Map<String, Object?>) throw const ParseException();
    return item;
  }).toList();
}

List<String> _strings(Object? value) {
  if (value is! List) throw const ParseException();
  return value.map(_text).toList();
}

File: backend/app/models/creator_pack.py
"""Audience-facing creator-pack contracts derived from an approved research map."""
from __future__ import annotations

from enum import StrEnum
from uuid import uuid4

from pydantic import BaseModel, Field, field_validator


class AudienceType(StrEnum):
    GENERAL_PUBLIC = "general_public"
    HIGH_SCHOOL = "high_school"
    UNDERGRADUATE = "undergraduate"


class CreatorPackStatus(StrEnum):
    DRAFT = "draft"
    APPROVED = "approved"


class EvidenceCard(BaseModel):
    finding_index: int = Field(ge=0)
    statement: str = Field(min_length=1)
    evidence_ids: list[str] = Field(min_length=1)
    confidence: str = Field(min_length=1)
    evidence: list["ResolvedEvidence"] = Field(default_factory=list)


class ResolvedEvidence(BaseModel):
    """Backend-resolved source record; never supplied by a model."""

    evidence_id: str = Field(min_length=1)
    chunk_id: str = Field(min_length=1)
    page: int = Field(ge=1)
    excerpt: str = Field(min_length=1, max_length=300)


class VisualAbstractBlock(BaseModel):
    label: str = Field(min_length=1)
    text: str = Field(min_length=1)
    evidence_ids: list[str] = Field(default_factory=list)


class GlossaryEntry(BaseModel):
    term: str = Field(min_length=1)
    definition: str = Field(min_length=1)


class CreatorPack(BaseModel):
    pack_id: str = Field(default_factory=lambda: str(uuid4()), min_length=1)
    paper_id: str = Field(min_length=1)
    audience: AudienceType
    status: CreatorPackStatus = CreatorPackStatus.DRAFT
    title: str = Field(min_length=1)
    summary: str = Field(min_length=1)
    narration_script: str = Field(min_length=1)
    visual_abstract: list[VisualAbstractBlock] = Field(min_length=1)
    evidence_cards: list[EvidenceCard] = Field(min_length=1)
    glossary: list[GlossaryEntry] = Field(default_factory=list)
    limitations: list[str] = Field(min_length=1)
    disclaimer: str = Field(min_length=1)

    @field_validator("paper_id", "title", "summary", "narration_script", "disclaimer", mode="before")
    @classmethod
    def _strip_text(cls, value: object) -> object:
        if isinstance(value, str):
            value = value.strip()
            if not value:
                raise ValueError("Field must not be blank")
        return value


class CreatorPackCreateRequest(BaseModel):
    audience: AudienceType


class CreatorPackUpdateRequest(BaseModel):
    title: str | None = None
    summary: str | None = None
    narration_script: str | None = None
    visual_abstract: list[VisualAbstractBlock] | None = None
    glossary: list[GlossaryEntry] | None = None


class CreatorPackApprovalRequest(BaseModel):
    approved: bool
